Tools/MiniBrowser/win/WebKitBrowserWindow.cpp

```diff
@@ -573,6 +578,64 @@ bool WebKitBrowserWindow::canTrustServerCertificate(WKProtectionSpaceRef protect
     return false;
 }
 
+void WebKitBrowserWindow::navigationActionDidBecomeDownload(WKPageRef, WKNavigationActionRef, WKDownloadRef download, const void* clientInfo)
+{
+    setDownloadClient(download, clientInfo);
+}
+
+void WebKitBrowserWindow::navigationResponseDidBecomeDownload(WKPageRef, WKNavigationResponseRef, WKDownloadRef download, const void* clientInfo)
+{
+    setDownloadClient(download, clientInfo);
+}
+
+void WebKitBrowserWindow::contextMenuDidCreateDownload(WKPageRef, WKDownloadRef download, const void* clientInfo)
+{
+    setDownloadClient(download, clientInfo);
+}
+
+void WebKitBrowserWindow::setDownloadClient(WKDownloadRef download, const void* clientInfo)
+{
+    WKDownloadClientV0 client = { };
+    client.base = { 0, clientInfo };
+    client.decideDestinationWithResponse = downloadDecideDestinationWithResponse;
+    client.didFinish = downloadDidFinish;
+    client.didFailWithError = downloadDidFailWithError;
+
+    WKDownloadSetClient(download, &client.base);
+}
+
+WKStringRef WebKitBrowserWindow::downloadDecideDestinationWithResponse(WKDownloadRef download, WKURLResponseRef response, WKStringRef suggestedFilename, const void* clientInfo)
+{
+    auto filename = createString(suggestedFilename);
+    if (filename.empty()) {
+        auto urlRef = adoptWK(WKURLResponseCopyURL(response));
+        auto hostRef = adoptWK(WKURLCopyHostName(urlRef.get()));
+        filename = createString(hostRef.get());
+    }
+
+    std::wstring folderPath;
+    getKnownFolderPath(FOLDERID_Downloads, folderPath);
```

Comment:
How should we handle an error of getKnownFolderPath? Can we cancel the download here?

Reply:
We can cancel the download by returning nullptr.

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/UIProcess/API/C/WKDownloadRef.cpp#L109-L113

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp#L177

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/NetworkProcess/NetworkProcess.cpp#L2122-L2123

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp#L312-L314

Reply:
Fixed